STYLE.md

```diff
@@ -0,0 +1,436 @@
+============================
+> This style guide is mostly a copy of [Johan Tibell's guide](https://github.com/tibbe/haskell-style-guide/blob/master/haskell-style.md)
+> with some restructurization, elaboration on some topics and some additions.
+> This style guide's aims are code beauty, readability and understandability.
+
+You can find our other formatting utilites and guidelines which expand the code style:
+* [.stylish-haskell.yaml](https://github.com/serokell/serokell-core/blob/master/HLint.hs)
+
+General guide lines
+-------------------
+### Line Length
+
+Maximum line length is *80 characters* or *100 characters* if necessary.
+
+Modern screens have high definition and big width.
+But with some tiling managers and two terminals on one screen you are not able to
+see many characters on one line.
+On the other hand, restricting line size to a very small number like 80 leads to
+some crazy indentation despite the fact that
+shorter lines should force you to write well structured code.
+That's why *100* is a reasonable compromize.
+
+
+### Indentation
+
+Tabs are illegal.  Use spaces for indenting.  Indent your code blocks
+with *2 spaces*.  Indent the `where` keyword with two spaces to set it
+apart from the rest of the code and indent the definitions in a
+`where` clause with 2 spaces.  Some examples:
+
+```haskell
+sayHello :: IO ()
+sayHello = do
+    name <- getLine
+    putStrLn $ greeting name
+  where
+    greeting name = "Hello, " ++ name ++ "!"
+
+filter :: (a -> Bool) -> [a] -> [a]
+filter _ [] = []
+filter p (x:xs)
+  | p x       = x : filter p xs
+  | otherwise = filter p xs
+```
+
+### Blank Lines
+
+* One blank line between top-level definitions.
+* No blank lines between type signatures and function definitions.
+* Add one blank line between
+functions in a type class instance declaration if the function bodies
+are large.
+* You can add blank lines inside a big `do` block to separate logical
+parts of it.
+* You can also use blank lines to separate definitions inside `where`
+clause, if necessary.
+
+### Whitespace
+
+* Surround binary operators with a single space on either side. In case of
+currying add one space between the argument and the operation.
+
+* Use some tool to remove trailing spaces.
+
+### Naming convention
+
+Casing:
++ **_lowerCamelCase_** for function and variable names.
++ **_UpperCamelCase_** for types.
+
+* Don't use short names like `n`, `sk`, `f` unless their meaning is clear from
+context (function name, types, other variables, etc.).
+
+* For readability reasons, don't capitalize all letters when using an
+abbreviation.  For example, write `HttpServer` instead of
+`HTTPServer`.  Exception: two or three letter abbreviations, e.g. `IO`, `STM`.
+
+**Records name conventions**
+
+If data type has only one constructor then this data type name should be same
+as constructor name (also applies to `newtype`).
+
+```haskell
+data User = User Int String
+```
+TODO: Convention for the `newtype` field name?
```

Comment:
@deepfire

Reply:
Do we need to mandate anything here?

I think that enforcing the `unThing` style often leads to overly long accessor names (which are OK for the newtype itself).

But on the other hand, not having to remember the name of the accessor is an upside.

Reply:
I'm happy to not mandate anything and just deal with on a case by case basis.

Reply:
Perhaps we could note it as a point to consider in the document.

Reply:
Updated.

Reply:
Specifically, I don't like `get${newtypeName}` because `get*` is so often used other places.

However, I think conventions are good and the larger the code base the higher the payoff. For the case of `newtype`s, I would suggest one of two options:

* No accessor so the user has to pattern match.
* Use `un${newtypeName}

For long lived, maintainable code the "overly long accessor names" problem is irrelevant. We should be writing code for long term readability not for being easy to write.

Reply:
> For long lived, maintainable code the "overly long accessor names" problem is irrelevant. We >should be writing code for long term readability not for being easy to write.

:100:

Reply:
> We should be writing code for long term readability not for being easy to write.

One would of course agree with that, but longer names do not necessarily translate to better readability -- simply because they use more space, so less code can be seen, and so because of the reduced density the big picture could become less accessible (forest, trees, etc.).

Generally, I'd cautiously agree on the `unThing` being a balanced approach, even despite the longer names.